include/dlaf/communication/functions.h

```diff
@@ -0,0 +1,76 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include <functional>
+
+#include "dlaf/communication/communicator.h"
+#include "dlaf/communication/mpi_datatypes.h"
+#include "dlaf/mpi_header.h"
+
+namespace dlaf {
+namespace comm {
+
+/// basic wrapper for MPI_Bcast
+template <typename MessageType>
+void bcast(int broadcaster_rank, MessageType& what, Communicator where) {
+  MPI_Bcast(&what, 1, mpi_datatype<MessageType>::type, broadcaster_rank, where);
+}
+
+/// basic wrapper for MPI_Ibcast
+template <typename MessageType>
+void async_bcast(int broadcaster_rank, MessageType& what, Communicator where,
+                 std::function<void()> action_before_retrying) {
+  MPI_Request request;
+  MPI_Ibcast(&what, 1, mpi_datatype<MessageType>::type, broadcaster_rank, where, &request);
+
+  while (true) {
+    int test_flag = 1;
+    MPI_Test(&request, &test_flag, MPI_STATUS_IGNORE);
+    if (test_flag)
+      break;
+    action_before_retrying();  // has this to be called from both sides of the communication?
+  }
+}
+
+namespace broadcast {
+
+/// specialized wrapper for MPI_Bcast on sender side
+template <typename MessageType>
+void send(const MessageType& what, Communicator where) {
+  bcast(where.rank(), const_cast<MessageType&>(what), where);
+}
+
+/// specialized wrapper for MPI_Bcast on receiver side
+template <typename MessageType>
+void receive_from(int broadcaster_rank, MessageType& what, Communicator where) {
+  bcast(broadcaster_rank, what, where);
+}
```

Comment:
Is a separate implementation between bcast send and bcast receive needed at this level?

Reply:
I have two arguments on this:

- The first one is that this "separated API" may be useful to have better "const-correctness" control over the message passed; in fact, the receive side requires a `non-const` message, while the sender side is ok both `const` and `non-const`. It is true that it hides a `const_cast` and from the MPI standard the constness is not guaranteed, since it says *"`MPI_BCAST` broadcasts a message from the process with rank root to all processes of the group, itself included. [...] On return, the content of root’s buffer is copied to all other processes.s"*;
- The second argument is that, looking at the prototype code, there are two different code sections for the sender and receiver side (mainly due to how the HPX taks has to be created).

Reply:
I agree with the first point.
For the second point I expect to have a front-end broadcast API that deals with tiles (see previous comment)